api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,244 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+  // The context class provides a context identifier. Is built as a linked list
+  // of context nodes which each have access to the nodes before it, but not 
+  // after. So each time a write is performed a new context is added to the list.
+  // Within each context node there is another linked list which holds the key
+  // and value data for each node.
+  class Context
+  {
+
+    public:
+
+
+      Context() = default;
+
+      // Creates a context object from a map of keys and identifiers, this will
+      // be the head of the context object linked list.
+      template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+        Context(const T &keys_and_values)  
+        {
+          head_ = new DataList(keys_and_values);
+        }
+
+      // Creates a context object from a key and value, this will be the head
+      // of the context object linked list.
+      Context(nostd::string_view key, ContextValue value)  
+      {
+        head_ = new DataList(key, value);
+      }
+
+      // Accepts a new iterable and then returns a new context that
+      // contains the new key and value data.
+      template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+        Context* SetValues(T &values) noexcept
+        {
+          return new Context(values, this);
+        }
+
+      // Accepts a new iterable and then returns a new context that
+      // contains the new key and value data.
+      Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+      {
+        return new Context(key, value, this);
+      }
+
+      // Returns the value associated with the passed in key.
+      common::AttributeValue GetValue(nostd::string_view key)
+      {
+        //Iterate through the context nodes
+        for(Context *context = this; context != nullptr; context = context->next_){
+          //Iterate through the internal data nodes
+          for(DataList *data = context->head_; data != nullptr; data = data->next_){
+            if(strncmp(key.data(), data->key_, data->key_length_) == 0){
```

Comment:
I think you want to first verify that the keys are the same length. Otherwise data's key could be a prefix of key (false positive) or you could overflow key.data() if it's shorter. 

```suggestion
            if(key.length() == data->key_length_ && strncmp(key.data(), data->key_, data->key_length_) == 0){
```